### PR TITLE
rgw, cls: add glue code and garbage collector for expired objects

### PR DESCRIPTION
**Pull request by rzarzynski**
Directly related with #4549 and #4719.
The last one was included here as a build dependency.

UPDATE:
To build the branch commit https://github.com/rzarzynski/ceph/commit/5c495f162f994feafba4e8f8c656fb86fc83ce82 from #4549 is necessary as well.

### CODE REVIEW

File: src/rgw/rgw_object_expirer.cc [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+  /* TODO: check whether the hint is actual. */
+  RGWObjectCtx rctx(store);
+
+  rgw_obj_key key = hint.obj_key;
+  if (key.instance.empty()) {
+    key.instance = "null";
+  }
+
+  rgw_obj obj(bucket_info.bucket, key);
+  ret = store->delete_obj(rctx, bucket_info, obj,
```
Comment (rzarzynski) [inline src/rgw/rgw_object_expirer.cc]: I think we need a call _set_atomic()_ here.

**Comment by yehudasa** [comment]
@rzarzynski I'll still need to go over it more thoroughly, but the idea is right. Also, as I mentioned earlier, the expirer should be split into core and main, so that we could use the core elsewhere.

**Comment by rzarzynski** [comment]
@yehudasa thank you for the info. It looks that I misunderstood you, sorry.
So do you want to have the expirer split into two modules: a really thin main (which contain the main loop, basic init and so on) and core (with functions for doing the real job like ```garbage_chunk```, ```trim_chunk```, ```inspect_all_shards``` etc.) for eventual reuse in the future. Am I right?

**Comment by rzarzynski** [comment]
It looks we got a merge conflict that hasn't been detected by github. Please take a look on [line 1718 in rgw_rados.cc](https://github.com/ceph/ceph/blob/master/src/rgw/rgw_rados.cc#L1718). I'm going to rebase the PR.

**Comment by rzarzynski** [comment]
@yehudasa this PR pull request has been rebased due to merge conflict. Only the [```2d3b10686e9fd32529c2aaf92b7fa3cc1dfe0d26```](https://github.com/rzarzynski/ceph/commit/2d3b10686e9fd32529c2aaf92b7fa3cc1dfe0d26) commit has been altered to reflect the changes in ```master```.

The [old branch](https://github.com/rzarzynski/ceph/tree/wip-4099_garbage_collector-old-4) is available as well.

**Comment by rzarzynski** [comment]
@yehudasa tested this branch altogether with #4549 against current master + PR #5047 + PR #5053 (small patches to prohibit early exit of many Tempest tests due to problems with _X-Trans-Id_ and _Content-Type_ HTTP headers). Both Tempest and s3-tests look fine.

* [Results of s3-tests and Tempest run on current Ceph master + PR #5047 + PR #5053](https://gist.github.com/rzarzynski/01d35481d9f7a0a1f052)
* [Results of s3-tests and Tempest run on current Ceph master + PR #5047 + PR #5053 + objexp (PR #4739 + PR #4549)](https://gist.github.com/rzarzynski/472d1ff03367990f2ddc)

I'm going to squash the small fixup on top of all other commits in this PR.

**Comment by rzarzynski** [comment]
Squashed the commit. Beside that, there is no other modifications. [Old branch](https://github.com/rzarzynski/ceph/tree/wip-4099_garbage_collector-old-5) is available for comparison.

**Comment by guangyy** [comment]
Hi @rzarzynski @yehudasa , where are we with this PR? We are interested in a S3' object lifecycle management feature and I think lots of ground work from this PR should help with the implementation of S3's lifecycle implementation.

**Comment by rzarzynski** [comment]
@yehudasa added support for CMake and rebased due to merge conflict.

[Previous branch state](https://github.com/rzarzynski/ceph/tree/wip-4099_garbage_collector-old-6) is available as well.

**Comment by yehudasa** [comment]
@guangyy still need to review it completely. I don't think there's much of a reuse here for S3, albeit we may consider extending our S3 to support this kind of object expiration. I think the main missing piece is that currently it requires running the cleanup process through radosgw-admin, and we probably want it to also happen similar to the GC process,

**Comment by rzarzynski** [comment]
@yehudasa currently GC of expired objects is provided through separated daemon called _radosgw-object-expirer_. As the most parts of code stay in form of library, we could very easily move the functionality to main rgw process, but we lose benefits from having address space separation in that way.

**Comment by rzarzynski** [comment]
@yehudasa rebased to merge conflict with current master.

**Comment by yehudasa** [comment]
@rzarzynski I don't see address space separation as a requirement for this feature, I think ease of use and deployment is far more important.

**Comment by yehudasa** [comment]
Opened PR #5692 against infernalis that also incorporates my additions

**Pull request closed without merging**